File: recette_talents_vues.js
/* Trois lectures d'un même relevé — et ce que chacune doit tenir.
 *
 * 1. LA SOUSTRACTION EXIGE SES DEUX TERMES. Quatre pays sur sept n'ont pas de
 *    lieu de travail publié : ils n'ont donc PAS de solde. Les poser à zéro
 *    les ferait passer pour équilibrés, ce qui est faux et flatteur — un zéro
 *    affiché serait une absence de calcul, pas une absence de mouvement. Ils
 *    doivent être nommés, avec le motif de leur absence.
 *
 * 2. UNE VALEUR NON PUBLIÉE NE SE DESSINE PAS COMME UNE VALEUR. Là où la
 *    source donne une fourchette (« 2–4 % »), la barre est hachurée et posée
 *    au milieu ; là où elle ne donne rien, il n'y a pas de barre du tout.
 *
 * 3. L'ÉCHELLE DIVERGENTE EST SYMÉTRIQUE. Un axe qui s'arrêterait à +29 d'un
 *    côté et à −14 de l'autre ferait paraître la perte chinoise plus grande
 *    qu'elle n'est. Zéro doit être au milieu, et le zéro doit se voir.
 *
 * 4. LES TROIS VUES DISENT LES MÊMES CHIFFRES. Trois formes, un seul relevé :
 *    si l'une d'elles s'écartait du tableau, c'est la page entière qui
 *    perdrait sa crédibilité.
 */
const { chromium } = require('/opt/node22/lib/node_modules/playwright');
const BASE = 'http://127.0.0.1:5401';
let ko = 0;
const ok = (n, c, d) => { console.log('  ' + (c ? 'OK ' : 'KO ') + '  ' + n + (d ? ' — ' + d : '')); if (!c) ko++; };

(async () => {
  const nav = await chromium.launch();
  const ctx = await nav.newContext({ viewport: { width: 1440, height: 1000 } });
  await ctx.addInitScript(() => {
    Object.defineProperty(navigator, 'webdriver', { get: () => false });
    Object.defineProperty(navigator, 'plugins', { get: () => [1, 2, 3] });
    Object.defineProperty(navigator, 'languages', { get: () => ['fr-FR', 'fr'] });
  });
  const pg = await ctx.newPage();
  const err = [];
  pg.on('pageerror', e => err.push(String(e)));
  await pg.goto(BASE + '/auth/recette_locale_idf_0123456789abcdef', { waitUntil: 'domcontentloaded' });
  await pg.goto(BASE + '/observatoire', { waitUntil: 'networkidle' });
  await pg.waitForFunction(() => document.querySelector('#tal-maps svg'), null, { timeout: 30000 });

  const vue = async (v) => {
    await pg.click('#tal-vues [data-tvue="' + v + '"]');
    await pg.waitForTimeout(400);
  };

  console.log('\n══ 1. Trois lectures, et chacune dit ce qu’elle tait ══\n');
  const onglets = await pg.evaluate(() => [...document.querySelectorAll('#tal-vues [data-tvue]')]
    .map(b => ({ v: b.getAttribute('data-tvue'), t: b.textContent.trim() })));
  ok('trois onglets', onglets.length === 3, onglets.map(o => o.v).join(','));
  ok('…le flux d’origine reste le premier, et par défaut',
     onglets[0].v === 'flux', onglets[0].t);
  ok('…suivi des deux lectures en barres',
     onglets[1].v === 'barres' && onglets[2].v === 'solde',
     onglets.map(o => o.t).join(' | '));

  const E = {};
  for (const v of ['flux', 'barres', 'solde']) {
    await vue(v);
    E[v] = await pg.evaluate(() => ({
      vue: TAL_VUE,
      titre: document.getElementById('t-tal').textContent,
      lecture: document.getElementById('tal-lecture').innerText,
      alt: document.querySelector('#tal-maps svg').getAttribute('aria-label'),
      selected: document.querySelector('#tal-vues [aria-selected="true"]').getAttribute('data-tvue'),
      legende: document.querySelector('#tal-maps .tal-leg').innerText,
    }));
  }
  ok('chaque onglet change la vue', Object.keys(E).every(v => E[v].vue === v));
  ok('…et l’état sélectionné suit pour le lecteur d’écran',
     Object.keys(E).every(v => E[v].selected === v));
  ok('trois titres de panneau distincts',
     new Set(Object.keys(E).map(v => E[v].titre)).size === 3,
     E.barres.titre + ' / ' + E.solde.titre);
  ok('…celui du solde annonce des POINTS, pas des pourcentages',
     /solde net.*en points/i.test(E.solde.titre), E.solde.titre);
  ok('trois textes de lecture distincts',
     new Set(Object.keys(E).map(v => E[v].lecture)).size === 3);
  ok('…et chacun écrit ce qu’il ne montre pas',
     Object.keys(E).every(v => /ne montre pas/i.test(E[v].lecture)));
  ok('trois descriptions alternatives distinctes et chiffrées',
     new Set(Object.keys(E).map(v => E[v].alt)).size === 3
     && Object.keys(E).every(v => /\d/.test(E[v].alt)));
  ok('chaque vue porte sa propre légende',
     new Set(Object.keys(E).map(v => E[v].legende)).size === 3);

  console.log('\n══ 2. Une valeur non publiée ne se dessine pas comme une valeur ══\n');
  await vue('barres');
  const bar = await pg.evaluate(() => {
    const svg = document.querySelector('#tal-maps svg');
    const pleines = [...svg.querySelectorAll('path[fill]')]
      .filter(p => /^#/.test(p.getAttribute('fill')));
    const hach = [...svg.querySelectorAll('path[fill^="url(#tal-hach"]')];
    const nd = [...svg.querySelectorAll('.tal-nd')].map(t => t.textContent.trim());
    const pays = DATA.talents.pays;
    return { rangs: svg.querySelectorAll('.tal-row').length,
             pleines: pleines.length, hachurees: hach.length, nd: nd,
             sansT: pays.filter(p => typeof p.t !== 'number').length,
             avecT: pays.filter(p => typeof p.t === 'number').length,
             fourchettes: pays.filter(p => /\d\s*[–-]\s*\d/.test(p.tn || '')).length,
             rien: pays.filter(p => typeof p.t !== 'number'
                                 && !/\d\s*[–-]\s*\d/.test(p.tn || '')).length };
  });
  ok('sept pays, sept lignes', bar.rangs === 7, bar.rangs);
  /* Une barre pleine par origine (7) + une par lieu de travail publié (3). */
  ok('dix barres pleines : sept origines, trois lieux de travail publiés',
     bar.pleines === bar.rangs + bar.avecT, bar.pleines + ' pour 7 + ' + bar.avecT);
  ok('…et trois barres HACHURÉES, une par fourchette publiée',
     bar.hachurees === bar.fourchettes && bar.fourchettes === 3,
     bar.hachurees + ' hachures pour ' + bar.fourchettes + ' fourchettes');
  ok('le pays sans aucune valeur n’a PAS de barre — juste la mention',
     bar.rien === 1 && bar.nd.indexOf('n.d.') >= 0, bar.nd.join(' · '));
  ok('…et chaque valeur non publiée est écrite en toutes lettres',
     bar.nd.length === bar.sansT, bar.nd.length + ' mentions pour ' + bar.sansT + ' inconnues');
  ok('la légende nomme la hachure', /valeur non publiée/i.test(E.barres.legende),
     E.barres.legende);
  /* DISCRIMINATION : la barre hachurée est posée au MILIEU de la fourchette,
     pas à son maximum — un maximum serait la lecture la plus flatteuse. */
  const milieu = await pg.evaluate(() => {
    const svg = document.querySelector('#tal-maps svg');
    const h = svg.querySelector('path[fill^="url(#tal-hach"]');
    const t = h.querySelector('title').textContent;
    return { titre: t, milieu: /au milieu de la fourchette/.test(t) };
  });
  ok('…et son infobulle dit qu’elle est posée au MILIEU de la fourchette',
     milieu.milieu, milieu.titre.slice(0, 80));

  console.log('\n══ 3. La soustraction exige ses deux termes ══\n');
  await vue('solde');
  const sol = await pg.evaluate(() => {
    const svg = document.querySelector('#tal-maps svg');
    const barres = [...svg.querySelectorAll('.tal-row')].map(g => ({
      pays: g.getAttribute('data-pays'),
      fill: g.querySelector('path[fill]').getAttribute('fill'),
      w: Math.round(g.querySelector('path[fill]').getBoundingClientRect().width),
      x: Math.round(g.querySelector('path[fill]').getBoundingClientRect().left) }));
    const pays = DATA.talents.pays;
    const calculables = pays.filter(p => talSolde(p) !== null);
    const absents = document.querySelector('#tal-maps p.note');
    return { barres: barres,
             calculables: calculables.length, total: pays.length,
             absents: absents ? absents.innerText : null,
             etiq: [...svg.querySelectorAll('.tal-sol')].map(t => t.textContent) };
  });
  ok('trois barres seulement, une par solde calculable',
     sol.barres.length === sol.calculables && sol.calculables === 3,
     sol.barres.length + ' barres pour ' + sol.calculables + ' soldes');
  /* LE contrôle : les quatre autres ne sont NULLE PART sur le graphique. */
  ok('…les quatre autres n’ont pas de barre nulle',
     sol.barres.length === 3 && sol.total === 7);
  ok('…mais ils sont NOMMÉS sous le graphique, avec leur motif',
     !!sol.absents && /Inde/.test(sol.absents) && /France/.test(sol.absents)
     && /Allemagne/.test(sol.absents) && /Canada/.test(sol.absents),
     (sol.absents || '').slice(0, 70));
  ok('…et le texte dit pourquoi : une soustraction exige ses deux termes',
     /soustraction exige ses deux termes/.test(sol.absents || ''));
  ok('…et refuse explicitement la barre nulle',
     /pas de barre plutôt qu’une barre nulle/.test(sol.absents || ''));
  ok('le compte est annoncé : 4 pays sur 7',
     /4 pays sur 7/.test(sol.absents || ''), (sol.absents || '').slice(0, 40));

  console.log('\n══ 4. L’axe divergent est symétrique, et son zéro se voit ══\n');
  const axe = await pg.evaluate(() => {
    const svg = document.querySelector('#tal-maps svg');
    const grad = [...svg.querySelectorAll('.tal-axe')].map(t => t.textContent.trim())
      .filter(t => /^[+−]?\d+$/.test(t));
    const zero = [...svg.querySelectorAll('.tal-grille')]
      .find(l => l.getAttribute('stroke') === '#8A8A8A');
    return { grad: grad, zeroMarque: !!zero,
             sens: [...svg.querySelectorAll('.tal-axe')].map(t => t.textContent.trim())
               .filter(t => /EXPORTE|ATTIRE/.test(t)) };
  });
  ok('les graduations vont de −30 à +30, symétriquement',
     axe.grad.join(' ') === '−30 −20 −10 0 +10 +20 +30', axe.grad.join(' '));
  /* Une échelle qui s'arrêterait à la valeur extrême de chaque côté ferait
     paraître −14 aussi long que +29. */
  ok('…et non de −14 à +29, ce qui égaliserait deux écarts inégaux',
     axe.grad.indexOf('−30') === 0 && axe.grad[axe.grad.length - 1] === '+30');
  ok('le zéro porte un trait plus marqué que les autres', axe.zeroMarque);
  ok('les deux côtés sont nommés', axe.sens.length === 2
     && axe.sens.join(' ').indexOf('EXPORTE') >= 0
     && axe.sens.join(' ').indexOf('ATTIRE') >= 0, axe.sens.join(' | '));
  ok('les longueurs suivent les valeurs : +29 est deux fois plus long que −14',
     Math.abs(sol.barres[0].w / sol.barres[2].w - 29 / 14) < 0.12,
     sol.barres[0].w + ' px vs ' + sol.barres[2].w + ' px');
  ok('la barre négative part bien à GAUCHE du zéro',
     sol.barres[2].x < sol.barres[0].x, sol.barres[2].x + ' < ' + sol.barres[0].x);
  ok('les couleurs opposent les deux sens',
     sol.barres[0].fill !== sol.barres[2].fill,
     sol.barres[0].fill + ' vs ' + sol.barres[2].fill);
  /* Une légende qui annonce un état qu'aucune barre n'occupe fait chercher au
     lecteur ce qui n'existe pas. */
  ok('la légende n’annonce pas l’état neutre, qu’aucune barre n’occupe',
     !/inférieur à 1,5/.test(E.solde.legende), E.solde.legende);

  console.log('\n══ 5. Le signe « moins » est le vrai, partout ══\n');
  const signes = await pg.evaluate(async () => {
    const lu = {};
    for (const v of ['flux', 'solde']) {
      document.querySelector('#tal-vues [data-tvue="' + v + '"]').click();
      await new Promise(r => setTimeout(r, 150));
      lu[v] = [...document.querySelectorAll('#tal-maps .tal-sol')].map(t => t.textContent.trim());
    }
    return lu;
  });
  /* U+2212 et non le trait d'union du clavier : le texte du panneau écrit
     « −14 points » et le graphique écrivait « -14 pts ». */
  ok('la vue solde écrit « −14 pts » avec un vrai moins',
     signes.solde.some(t => t.indexOf('−') === 0), signes.solde.join(' · '));
  ok('…et aucune étiquette ne porte le trait d’union',
     signes.solde.every(t => t.indexOf('-') < 0), signes.solde.join(' · '));
  ok('la vue flux a été alignée sur la même règle',
     signes.flux.some(t => t.indexOf('−') === 0)
     && signes.flux.every(t => t.indexOf('-') < 0), signes.flux.join(' · '));

  console.log('\n══ 6. Trois formes, un seul relevé ══\n');
  const memes = await pg.evaluate(async () => {
    const lu = {};
    for (const v of ['flux', 'barres', 'solde']) {
      document.querySelector('#tal-vues [data-tvue="' + v + '"]').click();
      await new Promise(r => setTimeout(r, 150));
      lu[v] = [...document.querySelectorAll('#tal-table tbody tr')]
        .map(tr => [...tr.children].map(c => c.textContent.trim()).join('|')).join(' ~ ');
    }
    /* L'ordre se relève sur la vue qui porte TOUS les pays : la vue solde n'en
       montre que trois, et y vérifier un classement complet ne prouverait
       rien. */
    document.querySelector('#tal-vues [data-tvue="barres"]').click();
    await new Promise(r => setTimeout(r, 150));
    const ordre = [...document.querySelectorAll('#tal-maps .tal-row')]
      .map(g => g.getAttribute('data-pays'));
    return { lu: lu, ordre: ordre };
  });
  ok('le tableau de données est le même sous les trois vues',
     memes.lu.flux === memes.lu.barres && memes.lu.barres === memes.lu.solde,
     memes.lu.solde.split(' ~ ')[0]);
  ok('…et il porte les sept pays', memes.lu.solde.split(' ~ ').length === 7);
  ok('les sept pays sont rangés du plus attractif au plus exportateur',
     memes.ordre.length === 7 && memes.ordre[0] === 'US' && memes.ordre[2] === 'CN',
     memes.ordre.join(','));
  /* Les soldes non calculables ferment la marche : ils ne se comparent à rien,
     et les glisser au milieu du classement leur prêterait un rang. */
  ok('…et les quatre sans solde ferment la marche',
     ['IN', 'FR', 'DE', 'CA'].every(c => memes.ordre.indexOf(c) >= 3),
     memes.ordre.slice(3).join(','));

  console.log('\n══ 7. Ce que les deux vues ne devaient PAS casser ══\n');
  await vue('flux');
  const flux = await pg.evaluate(() => {
    const svg = document.querySelector('#tal-maps svg');
    return { lignes: svg.querySelectorAll('.tal-row').length,
             fleches: svg.querySelectorAll('[marker-end]').length,
             creux: svg.querySelectorAll('circle[stroke-dasharray]').length,
             soldes: svg.querySelectorAll('.tal-sol').length };
  });
  ok('la figure de flux garde ses sept lignes', flux.lignes === 7, flux.lignes);
  ok('…ses flèches de sens', flux.fleches >= 2, flux.fleches);
  ok('…ses cercles creux pour les valeurs non publiées', flux.creux === 4, flux.creux);
  ok('…et sa colonne de soldes', flux.soldes === 7, flux.soldes);
  const survol = await pg.evaluate(() => {
    const g = document.querySelector('#tal-maps .tal-row');
    g.dispatchEvent(new MouseEvent('mouseenter', { bubbles: true, clientX: 300, clientY: 300 }));
    const t = document.querySelector('.tip, #tip, .obs-tip');
    return t ? t.innerText.slice(0, 60) : null;
  });
  ok('le survol répond toujours', !!survol, survol);
  ok('aucune erreur JavaScript', err.length === 0, err.slice(0, 2).join(' | '));

  await nav.close();
  console.log('');
  console.log(ko ? ko + ' contrôle(s) en échec\n' : 'tout est vert\n');
  process.exit(ko ? 1 : 0);
})();

File: recette_talents_vues.py
# -*- coding: utf-8 -*-
"""Trois lectures des chercheurs d'élite — côté données et côté dépôt.

CE QU'ON PROTÈGE.

1. LE SOLDE N'EXISTE QUE SI SES DEUX TERMES EXISTENT. Quatre pays sur sept
   n'ont pas de lieu de travail publié. Les compter à zéro les ferait passer
   pour équilibrés — la lecture la plus flatteuse, et la plus fausse. Le
   contrôle rejoue le calcul refusé.

2. UNE FOURCHETTE N'EST PAS UNE VALEUR. « 2–4 % » sert à POSER un repère au
   milieu, jamais à calculer un solde. On vérifie que les deux usages restent
   séparés dans le code.

3. UNE SEULE ÉCRITURE DU SIGNE MOINS. Le texte du panneau écrivait « −14 »
   (U+2212) et le graphique « -14 » (trait d'union) : deux glyphes pour la même
   chose sur le même écran.

4. LA DISCRIMINATION. Avant, le panneau n'avait aucune bascule de vue, et
   `renderTalents` faisait DEUX choses sous un seul nom — la figure et le
   tableau. Ajouter une bascule du même nom aurait écrasé la fonction
   d'origine sans bruit.
"""
import io
import os
import re
import subprocess
import sys

DEPOT = "/home/user/conseilprev"
sys.path.insert(0, DEPOT)
os.chdir(DEPOT)

ko = 0


def ok(nom, cond, detail=""):
    global ko
    print("  %s   %s%s" % ("OK " if cond else "KO ", nom,
                           (" — " + str(detail)) if detail else ""))
    if not cond:
        ko += 1


import observatoire_ia as O                                    # noqa: E402

T = O.SEED["talents"]
ORI = T["origine_pct"]
TRA = T["lieu_travail_pct"]
html = io.open(DEPOT + "/observatoire.html", encoding="utf-8").read()

print("\n══ 1. Le solde n'existe que si ses deux termes existent ══\n")

ok("sept origines publiées", len(ORI) == 7, sorted(ORI))
ok("…et trois lieux de travail seulement", len(TRA) == 3, sorted(TRA))
calculables = [p for p in ORI if p in TRA]
ok("le solde n'est donc calculable que pour trois pays",
   len(calculables) == 3, sorted(calculables))
ok("…États-Unis +29 points", TRA["États-Unis"] - ORI["États-Unis"] == 29)
ok("…Chine −14 points", TRA["Chine"] - ORI["Chine"] == -14)
ok("…et « Autres » +3 points", TRA["Autres"] - ORI["Autres"] == 3)
# DISCRIMINATION : c'est le calcul qu'on refuse d'écrire. À zéro, l'Inde
# passerait pour un pays qui emploie exactement ce qu'il forme.
muets = [p for p in ORI if p not in TRA]
ok("quatre pays n'ont pas de solde", len(muets) == 4, sorted(muets))
ok("…et les poser à zéro les dirait « à l'équilibre », ce qu'aucune source "
   "n'affirme", all(ORI[p] > 0 for p in muets),
   {p: ORI[p] for p in sorted(muets)})
ok("l'Inde en fait partie, avec 7 % de chercheurs formés",
   "Inde" in muets and ORI["Inde"] == 7)

print("\n══ 2. Une fourchette n'est pas une valeur ══\n")

ok("la note dit que quatre pays sont entre 2 et 4 %",
   "entre 2 et 4 %" in T["note_lieu_travail"])
ok("…et que la valeur précise n'est pas publiée",
   "non publiées" in T["note_lieu_travail"])
ok("la page lit la fourchette par une fonction dédiée",
   html.count("function talFourchette") == 1)
ok("…qui rend un intervalle, jamais un nombre",
   "return m ? [+m[1], +m[2]] : null;" in html)
ok("…et la barre hachurée se pose au MILIEU, pas au maximum",
   "var mil = (f[0] + f[1]) / 2;" in html)
ok("le solde, lui, exige un nombre : `typeof p.t !== 'number'` rend null",
   "if(typeof p.o !== 'number' || typeof p.t !== 'number') return null;" in html)
# La fourchette ne doit JAMAIS entrer dans talSolde : ce serait un solde
# fabriqué à partir d'une lecture graphique.
bloc = html[html.find("function talSolde(p){"):]
bloc = bloc[:bloc.find("\n}")]
ok("…et `talSolde` ignore totalement la fourchette",
   "talFourchette" not in bloc and "tn" not in bloc, bloc[-80:])

print("\n══ 3. Un seul signe moins, et une seule fonction pour l'écrire ══\n")

ok("la fonction d'écriture des points existe", "function talPts(" in html)
ok("…et elle emploie le VRAI moins (U+2212)", "'−'" in html)
ok("…appelée par les trois vues et l'axe", html.count("talPts(") >= 5,
   html.count("talPts("))
ok("plus aucune étiquette ne fabrique son signe à la main",
   "(sol > 0 ? '+' : '') + fr(Math.round(sol" not in html)
ok("le texte du panneau écrivait déjà ce moins-là", "−14 points" in html)

print("\n══ 4. Trois lectures déclarées, et servies ══\n")

for cle in ("flux", "barres", "solde"):
    ok("l'onglet « %s » existe" % cle, ('data-tvue="%s"' % cle) in html)
ok("trois titres, un par lecture",
   len(re.findall(r"^\s+(flux|barres|solde):\s+'Chercheurs", html, re.M)) == 3)
ok("les deux nouvelles fonctions de rendu existent",
   "function talBarres()" in html and "function talDiverge()" in html)
ok("…et le dispatch leur passe la main",
   "if(TAL_VUE === 'barres') return talBarres();" in html)
ok("les trois textes de lecture disent ce qu'ils ne montrent pas",
   html.count("Ce qu’elle ne montre pas") >= 3
   or html.count("ne montre pas") >= 3, html.count("ne montre pas"))
ok("la barre horizontale a son propre tracé, arrondi côté valeur",
   "function barPathH(" in html and "function barPathHG(" in html)
ok("…et le miroir gauche existe pour les soldes négatifs",
   "arrondie du côté GAUCHE" in html)
ok("l'échelle divergente est symétrique, ±30",
   "var VMAX = 30;" in html and "[-30, -20, -10, 0, 10, 20, 30]" in html)

print("\n══ 5. Discrimination : ni bascule, ni barres, avant ══\n")


def _avant(marqueur, fichier):
    """Le fichier juste AVANT l'arrivée du marqueur.

    Si un commit l'a introduit, on lit son parent ; sinon la modification n'est
    pas encore livrée et la référence est HEAD. Rendre une chaîne vide rendrait
    le contrôle CREUX."""
    hs = subprocess.check_output(
        ["git", "-C", DEPOT, "log", "-S", marqueur, "--format=%H", "--", fichier],
        text=True).split()
    ref = ("%s^" % hs[-1]) if hs else "HEAD"
    return subprocess.check_output(
        ["git", "-C", DEPOT, "show", "%s:%s" % (ref, fichier)], text=True)


av = _avant("function talBarres", "observatoire.html")
ok("avant, le panneau n'avait aucune bascule de vue", "data-tvue" not in av)
ok("…ni barres jumelées", "talBarres" not in av)
ok("…ni barres divergentes", "talDiverge" not in av)
ok("…et une seule figure, le flux", av.count("function renderFluxTalents") == 1)
# LE piège : `renderTalents` faisait DEUX choses sous un seul nom. Ajouter une
# bascule du meme nom aurait écrasé l'original — et c'est la derniere
# définition qui l'emporte, en silence.
ok("…tandis que `renderTalents` faisait AUSSI le tableau",
   "function renderTalents(){\n  renderFluxTalents();" in av)
ok("le tableau a donc été détaché sous son propre nom",
   "function renderTalTable()" in html
   and "function renderTalents(){\n  renderFluxTalents();" not in html)
ok("…et il n'existe qu'UNE définition de renderTalents",
   html.count("function renderTalents(") == 1, html.count("function renderTalents("))
ok("avant, l'étiquette de solde fabriquait son signe sur place",
   "(sol > 0 ? '+' : '') + fr(Math.round(sol * 10) / 10) + ' pts'" in av)

print("\n══ 6. Ce que ces deux lectures ne devaient PAS déplacer ══\n")

ok("les pourcentages publiés sont intacts",
   ORI["États-Unis"] == 28 and ORI["Chine"] == 26 and TRA["États-Unis"] == 57
   and TRA["Chine"] == 12)
ok("la définition « présentation orale » est toujours là",
   "présentation ORALE" in T["definition"] or "orale" in T["definition"])
ok("le crédit MacroPolo est intact", "MacroPolo" in T["source"])
ok("…et la mention de citation obligatoire", "citation" in T["licence"])
ok("la précision distingue toujours l'exact du lu sur carte",
   "lecture graphique" in T["precision"] and "publiés" in T["precision"])
# Le compte porte sur les APPELS : la ligne de définition contient elle aussi
# « talSurvol(el » et gonflait le total d'un.
ok("le survol est partagé par les trois vues, en une fonction",
   html.count("function talSurvol(") == 1 and html.count("  talSurvol(el") == 3,
   html.count("  talSurvol(el"))
ok("le tri est partagé lui aussi", html.count("function talListe(") == 1)

print("")
print("%d contrôle(s) en échec\n" % ko if ko else "tout est vert\n")
sys.exit(1 if ko else 0)
